### PR TITLE
[X86] Support "f16c" and "avx512fp16" for __builtin_cpu_supports

### PR DESCRIPTION
**Pull request by FreddyLeaf**
This resolves issue #65320.
This also supports clarify sapphirerapids and cooperlake for cpu_specific/dispatch.

### CODE REVIEW

File: llvm/include/llvm/TargetParser/X86TargetParser.def [hunk 1]
```diff
@@ -229,9 +230,9 @@ X86_FEATURE       (XSAVE,           "xsave")
 X86_FEATURE       (XSAVEC,          "xsavec")
 X86_FEATURE       (XSAVEOPT,        "xsaveopt")
 X86_FEATURE       (XSAVES,          "xsaves")
+X86_FEATURE_COMPAT(AVX512FP16,      "avx512fp16",            39)
```
Comment (phoebewang) [inline llvm/include/llvm/TargetParser/X86TargetParser.def:233]: Can we add this feature as key feature to sapphirerapids now?

File: llvm/include/llvm/TargetParser/X86TargetParser.def [hunk 2]
```diff
@@ -184,12 +185,12 @@ X86_FEATURE       (AMX_TILE,        "amx-tile")
 X86_FEATURE       (CLDEMOTE,        "cldemote")
 X86_FEATURE       (CLFLUSHOPT,      "clflushopt")
 X86_FEATURE       (CLWB,            "clwb")
+X86_FEATURE_COMPAT(F16C,            "f16c",                  38)
```
Comment (phoebewang) [inline llvm/include/llvm/TargetParser/X86TargetParser.def:188]: Can we test them in clang/test?

File: llvm/include/llvm/TargetParser/X86TargetParser.def (hunk 1 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/TargetParser/X86TargetParser.def:233]: You mean update getIntelProcessorTypeAndSubtype, right? I think we don't need to do after we already updated the model value for sapphirerapids .

File: llvm/include/llvm/TargetParser/X86TargetParser.def (hunk 2 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/TargetParser/X86TargetParser.def:188]: agree, will do

Comment (FreddyLeaf) [reply llvm/include/llvm/TargetParser/X86TargetParser.def:188]: 79bf68d

File: llvm/include/llvm/TargetParser/X86TargetParser.def (hunk 1 above)
Comment (phoebewang) [reply llvm/include/llvm/TargetParser/X86TargetParser.def:233]: https://github.com/llvm/llvm-project/blob/main/llvm/lib/TargetParser/X86TargetParser.cpp#L350
I recall we are different with GCC due to the support of AVX512FP16.

File: clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp [hunk 3]
```diff
@@ -80,17 +80,17 @@ OutOfLineDefs::foo(int, int, int) {
 // LINUX: define dso_local noundef i32 @_ZN13OutOfLineDefs3fooEiii.S
 // LINUX: define dso_local noundef i32 @_ZN13OutOfLineDefs3fooEiii.R
 // LINUX: define weak_odr ptr @_ZN13OutOfLineDefs3fooEiii.resolver()
-// LINUX: ret ptr @_ZN13OutOfLineDefs3fooEiii.R
 // LINUX: ret ptr @_ZN13OutOfLineDefs3fooEiii.S
+// LINUX: ret ptr @_ZN13OutOfLineDefs3fooEiii.R
```
Comment (phoebewang) [inline clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp:84]: Is it affected by `f16c` feature? I'm worried if the test will fail with some old GCC version.

File: llvm/include/llvm/TargetParser/X86TargetParser.def (hunk 1 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/TargetParser/X86TargetParser.def:233]: Sorry, I misunderstood before. c50fb21 updated. With this update, cpu_specific/dispatch can clarify sapphirerapids and cooperlake now.

File: clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp (hunk 3 above)
Comment (FreddyLeaf) [reply clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp:84]: Yes, I guess it will. Need some environment to verify.

Comment (FreddyLeaf) [reply clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp:84]: We landed https://reviews.llvm.org/D158811 before, I think this won't be blocking issue? 
CC @MaskRay

**Review by phoebewang — APPROVED** [review]
LGTM.

File: clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp (hunk 3 above)
Comment (MaskRay) [reply clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp:84]: Reordering `X86_FEATURE` causes the change here. This should not matter. Ideally, the test can be improved to be check the relevant `__cpu_model` members.

**Review by MaskRay — APPROVED** [review]

File: clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp (hunk 3 above)
Comment (FreddyLeaf) [reply clang/test/CodeGenCXX/attr-cpuspecific-outoflinedefs.cpp:84]: Thanks comment!

**Pull request merged**